docs/tradeoffs.md

```diff
@@ -0,0 +1,136 @@
+[bpaf](https://docs.rs/bpaf) and [clap](https://docs.rs/clap).
+
+For anyone asking the question "which should I use?", the short answer would be
+- `bpaf` would work well for simple cases like example bins for a library
+  while giving the best build times and a reasonable to understand code with
+  the `derive` API
+- All other cases the answer is "it depends" as the two designs offer different
+  trade offs.   If you want to just choose one with little research that will
+  cover the most use cases, that will most likely be `clap`.
+
+## Static vs Dynamic Typing
+
+`bpaf`'s combinator API uses generics and macros to build a statically-typed
+parser, pushing development errors to compile time.  This minimizes binary
+size and runtime, only paying for what you use.
+
+The combinator approach tends towards yoda-speak though with [careful
+structuring by breaking down arguments into
+functions](https://github.com/pacak/bpaf/blob/aa6992931bbfbdca6390c87f4a76898f8db0ae47/examples/top_to_bottom.rs),
+a more straightforward ordering can be accomplished.
+
+`clap`'s builder API stores parsed values in a Map-like container where the
+keys are the argument IDs and the values are effectively `Box<dyn Any>`.
+- This allows dynamically created CLIs, like with
+  [clap_serde](https://docs.rs/clap_serde) or [conditionally-present
+  flags](https://github.com/sharkdp/bat/blob/6680f65e4b25b0f18c455f7a4639a96e97519dc5/src/bin/bat/clap_app.rs#L556)
+- Callers can dynamically iterate over the arguments, like for layered configs.
+- `clap` can have ready-to-use, arbitrarily deifned defaults and validation
+  conditioned on argument values (e.g. `default_value_if`).  See the section on
+  Validation for more nuance on the "arbitrarily" part
+- `clap` tries to help push as many errors to early test time with minimal
+  effort with the `cmd.debug_assert()` function.  Value lookups still requires
+  full coverage of that code to catch errors.
+
+Which design approach is faster to build will be dependent on the exact
+implementation and the compiler.
+
+## Context-sensitive parsing
+
+Parsing arguments is context sensitive.  Consider:
+```console
+$ prog --one two --three --four -abcdef
+```
+- When is an argument a subcommand, a positional value, a flag, or a flag's value?
+- When is the trailing parts of a short flag an attached value or more shorts flags?
+
+For example, some possible interpretations of the above could be:
+```console
+$ prog --one=two --three=--four -a=bcdef
+$ prog two -a -b -c -d -e -f --one --three --four
+```
+
+`clap` parses left-to-right using [`clap_lex`](https://docs.rs/clap_lex), with
+the output from the previous token hinting how to parse the next one to
+prevent ambiguity.  This leads to a fairly complicated parser to handle all of
+the different cases upfront, including some usability features to help catch
+developer mistakes.
+
+`bpaf` instead does context-free tokenization, storing all flags and values in
```

Comment:
Tokenizer still produces a list - not having it being coupled with the parser makes things easier, decision to how to consume it is made during parsing specifically. As a side effect this allows to use `-o` as both boolean switch and an argument for `-o foo.rs`/`-ofoo.rs`, and only when the usage is ambiguous it would complain to user to rephrase it: `-ov` with both `-o` and `-v` flags and `-o` argument present => ask user to specify either `-o -v` or `-o=v`.